File: sdk-rust/examples/describe-image.rs
use base64::{engine::general_purpose::STANDARD as BASE64_STANDARD, Engine};
use dotenvy::dotenv;
use llm_sdk::{LanguageModelInput, Message, Part};

mod common;

#[tokio::main]
async fn main() {
    dotenv().ok();

    let image_url = "https://images.unsplash.com/photo-1464809142576-df63ca4ed7f0";
    let image_res = reqwest::get(image_url)
        .await
        .expect("failed to fetch image");
    let mime_type = image_res
        .headers()
        .get(reqwest::header::CONTENT_TYPE)
        .and_then(|ct| ct.to_str().ok())
        .unwrap_or("image/jpeg")
        .to_string();
    let image_bytes = image_res.bytes().await.expect("failed to read bytes");

    let image_b64 = BASE64_STANDARD.encode(&image_bytes);

    let model = common::get_model("openai", "gpt-4o");

    let response = model
        .generate(LanguageModelInput {
            messages: vec![Message::user(vec![
                Part::text("Describe this image"),
                Part::image(image_b64, mime_type),
            ])],
            ..Default::default()
        })
        .await
        .expect("model.generate failed");

    println!("{response:#?}");
}

File: sdk-js/examples/describe-image.ts
import { getModel } from "./get-model.ts";

const imageUrl = "https://images.unsplash.com/photo-1464809142576-df63ca4ed7f0";
const imageRes = await fetch(imageUrl);

const image = await imageRes.arrayBuffer();

const model = getModel("openai", "gpt-4o");

const response = await model.generate({
  messages: [
    {
      role: "user",
      content: [
        {
          type: "text",
          text: "Describe this image",
        },
        {
          type: "image",
          data: Buffer.from(image).toString("base64"),
          mime_type: imageRes.headers.get("content-type") ?? "image/jpeg",
        },
      ],
    },
  ],
});

console.dir(response, { depth: null });
